{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "55ad06d4-b76b-4f6f-88b5-6a4db236739c",
   "metadata": {},
   "source": [
    "# Day 4 Part 1\n",
    "---\n",
    "\n",
    "## Multiple sequence aligners + Phylogenetics\n",
    "\n",
    "For more information about the theory behind of multiple sequence alignment, please see these resources:\n",
    "\n",
    "- https://academic.oup.com/bib/article/17/6/1009/2606431\n",
    "- https://www.hindawi.com/journals/isrn/2013/615630/\n",
    "\n",
    "Note: There is a user-friend tool called MEGA that is likely sufficient for analyzing closely related species (within a phylum). However, it lacks some of the more sophisticated phylogenetic methods https://www.megasoftware.net/\n",
    "\n",
    "### Aligners\n",
    "\n",
    "Here is a list of recommended aligners (non-exhaustive). Unfortunately, most do not have Windows packages, however some of them do have webservers.  \n",
    "- [MAFFT](https://mafft.cbrc.jp/alignment/software/) \n",
    "- [T-COFFEE](http://www.tcoffee.org/Projects/tcoffee/workshops/tcoffeetutorials/installation.html)\n",
    "- [MUSCLE](https://www.drive5.com/muscle/) -- Has a windows distribution!\n",
    "- [HMMER](http://hmmer.org/) *great for aligning distantly related proteins*\n",
    "  - Profile alignment, guides the alignment of your sequences using a well curated template file. \n",
    "\n",
    "### Alignment viewers\n",
    "\n",
    "The most popular viewer on the market is [Jalview](https://www.jalview.org/), but there are older alternatives such as [SeaView](http://doua.prabi.fr/software/seaview).\n",
    "\n",
    "[Windows](https://www.jalview.org/download/windows/)\n",
    "[Mac](https://www.jalview.org/download/macos/)\n",
    "[Linux](https://www.jalview.org/download/linux/)\n",
    "\n",
    "\n",
    "### Masking alignments\n",
    "\n",
    "If you want to proceed to phylogenetic analysis you can treat your alignment with a 'masking' method. This is a way to 'mask' or 'hide' regions of the alignment that might be misaligned (and therefore non-homologous). Historically, this was done by eye (yep..) but now we have tools to help us out! My favourites are: \n",
    "\n",
    "- [BMGE](https://bmcecolevol.biomedcentral.com/articles/10.1186/1471-2148-10-210)\n",
    "- [trimal](http://trimal.cgenomics.org/) -- has a windows distribution\n",
    "- [divvier](https://github.com/simonwhelan/Divvier) # advanced user\n",
    "\n",
    "### Phylogenetic Analyses\n",
    "\n",
    "Finally, for phylogenetic analyses, I recommend Maximum-likelihood or Bayesian methods:\n",
    "\n",
    "- [IQTREE](http://www.iqtree.org/)\n",
    "- [RAxML](https://cme.h-its.org/exelixis/web/software/raxml/)\n",
    "- [PhyloBayes](http://www.atgc-montpellier.fr/phylobayes/)\n",
    "\n",
    "### TreeViewers\n",
    "\n",
    "- [FIGTREE](http://tree.bio.ed.ac.uk/software/figtree/)\n",
    "- [iTOL](https://itol.embl.de/)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7bb5a53-7f33-4e09-a7a2-cdfbbcf2a226",
   "metadata": {},
   "source": [
    "Since muscle has a windows distribution let's try and get that working.  Mac/unix/colab users head to the next section\n",
    "\n",
    "### Windows\n",
    "\n",
    "Window's folks head to this link and download the executable (tar.gz again): \n",
    "\n",
    "https://www.drive5.com/muscle/downloads.htm\n",
    "\n",
    "http://trimal.cgenomics.org/downloads\n",
    "\n",
    "http://www.iqtree.org/#download\n",
    "\n",
    "\n",
    "Note the path where you have downlodaded these including the \\bin\\\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0159a466-17cf-4bcb-b820-faf13338ca84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Jalview installation \n",
    "\n",
    "# Install Jalview following the instructions above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06f2b420-be4f-4554-afdb-c452095e9bd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Muscle installation\n",
    "\n",
    "# Windows/Mac \n",
    "# follow instructions\n",
    "\n",
    "# windows path setting\n",
    "!set PATH=%PATH%;C:\\your\\path\\here\\\n",
    "\n",
    "# mac path setting\n",
    "export PATH=$PATH:/your/path/here\n",
    "\n",
    "# Linux\n",
    "# download file\n",
    "# make it executable chmod 755 file\n",
    "# run the file ./file\n",
    "# not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "782091cc-cf7a-474f-8691-3cfdd185ae82",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "muscle 5.1.linux64 [12f0e2]  15.8Gb RAM, 8 cores\n",
      "Built Jan 13 2022 23:17:13\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 86 seqs, avg length 577, max 714\n",
      "\n",
      "00:00 4.4Mb  CPU has 8 cores, running 8 threads\n",
      "00:27 559Mb   100.0% Calc posteriors\n",
      "00:31 559Mb   100.0% Consistency (1/2)\n",
      "00:35 559Mb   100.0% Consistency (2/2)\n",
      "00:35 559Mb   100.0% UPGMA5           \n",
      "00:36 571Mb   100.0% Refining\n"
     ]
    }
   ],
   "source": [
    "## Run the aligner using muscle\n",
    "## It will save to the file SDHA_ncbi_align.fasta \n",
    "\n",
    "import subprocess\n",
    "muscle = \"/home/julieb/SOFTS/muscle5.1.linux_intel64\" # write here the path of your muscle install\n",
    "in_file = \"SDHA_ncbi.fasta\"\n",
    "out_file = \"SDHA_ncbi_muscle.fasta\"\n",
    "muscle_result=subprocess.run([muscle, \"-align\", in_file, \"-output\", out_file])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "180baeb1-4f5b-4df0-9a9e-bdecde449c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open from a console or by clicking on the icon shortcut\n",
    "# jalview "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9726eb2b-25fc-43b6-ab12-0e0bdc3ab09d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### windows\n",
    "# should be like muscle \n",
    "\n",
    "### MAC/UNIX/COLLAB\n",
    "!conda install -c bioconda --yes trimal\n",
    "!conda install --yes -c bioconda iqtree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "53833254-f628-4cff-8e2d-f5ae9d9a0d55",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run the 'masking' or 'trimming' tool\n",
    "\n",
    "!trimal -gappyout -in SDHA_ncbi_muscle.fasta  -out SDHA_ncbi_muscle_trim.fasta \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "666e2c98-e62a-417e-a71b-f1974f3f33c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## if it does not work, try \n",
    "\n",
    "trimal = \"/home/julieb/anaconda3/bin/trimal\" # write here the path of your muscle install\n",
    "in_file = \"SDHA_ncbi_muscle.fasta\"\n",
    "out_file = \"SDHA_ncbi_muscle_trim2.fasta\"\n",
    "trimal_result=subprocess.run([trimal, \"-gappyout\", \"-in\", in_file, \"-out\", out_file])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fb3e582-cfb3-4a59-90ea-37fbbe78d2ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# then look at your alignment in jalview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "24866bba-69e7-45b8-beeb-d6d6c9f25784",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.0.3 for Linux 64-bit built Dec 20 2020\n",
      "Developed by Bui Quang Minh, Nguyen Lam Tung, Olga Chernomor,\n",
      "Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    julieb-Latitude-5420 (AVX512, FMA3, 15 GB RAM)\n",
      "Command: iqtree -s SDHA_ncbi_muscle_trim.fasta -m LG -B 1000 -nt 6 -pre SDHA_ncbi_muscle_trim\n",
      "Seed:    393519 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Thu Sep 29 11:10:01 2022\n",
      "Kernel:  AVX+FMA - 6 threads (8 CPU cores detected)\n",
      "\n",
      "Reading alignment file SDHA_ncbi_muscle_trim.fasta ... Fasta format detected\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 86 sequences with 441 columns, 424 distinct patterns\n",
      "404 parsimony-informative, 11 singleton sites, 26 constant sites\n",
      "WARNING: Some sequence names are changed as follows:\n",
      "Q8YXJ6.1_Nostoc_sp._PCC_7120_=_FACHB-418 -> Q8YXJ6.1_Nostoc_sp._PCC_7120___FACHB-418\n",
      "Q9X8N8.1_Streptomyces_coelicolor_A3(2) -> Q9X8N8.1_Streptomyces_coelicolor_A3_2_\n",
      "\n",
      "                                                                           Gap/Ambiguity  Composition  p-value\n",
      "   1  Q6ZDY8.1_Oryza_sativa_Japonica_Group                                         0.00%    passed     95.93%\n",
      "   2  P31040.2_Homo_sapiens                                                        0.00%    passed     76.00%\n",
      "   3  P17412.3_Wolinella_succinogenes_DSM_1740                                     0.00%    passed     42.55%\n",
      "   4  Q8TZL4.1_Pyrococcus_furiosus_DSM_3638                                        1.81%    failed      0.00%\n",
      "   5  Q8YXJ6.1_Nostoc_sp._PCC_7120___FACHB-418                                     0.00%    failed      2.29%\n",
      "   6  Q8XNE2.1_Clostridium_perfringens_str._13                                     7.26%    failed      0.00%\n",
      "   7  P31038.1_Rickettsia_prowazekii_str._Madrid_E                                 0.00%    passed    100.00%\n",
      "   8  Q92J97.1_Rickettsia_conorii_str._Malish_7                                    0.00%    passed     99.90%\n",
      "   9  Q9JSX4.1_Neisseria_meningitidis_Z2491                                        0.68%    failed      0.01%\n",
      "  10  P10902.4_Escherichia_coli_K-12                                               0.00%    passed     88.02%\n",
      "  11  Q9X8N8.1_Streptomyces_coelicolor_A3_2_                                       0.00%    failed      0.00%\n",
      "  12  Q6PA58.1_Xenopus_laevis                                                      0.00%    passed     90.85%\n",
      "  13  Q8Y5N4.1_Listeria_monocytogenes_EGD-e                                        0.91%    failed      0.06%\n",
      "  14  Q6Z836.1_Oryza_sativa_Japonica_Group                                         0.00%    passed     93.27%\n",
      "  15  Q68XN9.1_Rickettsia_typhi_str._Wilmington                                    0.00%    passed     99.99%\n",
      "  16  Q92R32.1_Sinorhizobium_meliloti_1021                                         0.45%    passed      5.24%\n",
      "  17  O57765.1_Pyrococcus_horikoshii_OT3                                           1.36%    failed      0.03%\n",
      "  18  Q9HNZ0.1_Halobacterium_salinarum_NRC-1                                       1.36%    failed      0.00%\n",
      "  19  Q920L2.1_Rattus_norvegicus                                                   0.00%    passed     65.08%\n",
      "  20  V3TQ67.1_Serratia_sp._ATCC_39006                                             0.00%    passed     69.88%\n",
      "  21  Q4UJM1.1_Rickettsia_felis_URRWXCal2                                          0.00%    passed     99.89%\n",
      "  22  P38032.1_Bacillus_subtilis_subsp._subtilis_str._168                          0.68%    passed     69.80%\n",
      "  23  Q9PC57.1_Xylella_fastidiosa_9a5c                                             0.23%    failed      0.20%\n",
      "  24  Q87D19.1_Xylella_fastidiosa_Temecula1                                        0.23%    failed      0.13%\n",
      "  25  Q97ZC5.2_Saccharolobus_solfataricus_P2                                       3.17%    failed      0.00%\n",
      "  26  P0C278.1_Shewanella_frigidimarina                                           16.33%    failed      0.00%\n",
      "  27  O82663.1_Arabidopsis_thaliana                                                0.00%    passed     99.53%\n",
      "  28  Q9V2R0.1_Pyrococcus_abyssi_GE5                                               1.36%    failed      0.02%\n",
      "  29  Q8XQG4.1_Ralstonia_solanacearum_GMI1000                                      0.23%    passed     85.97%\n",
      "  30  Q0QF17.1_Mesocricetus_auratus                                                0.68%    passed     76.29%\n",
      "  31  Q929Z2.1_Listeria_innocua_Clip11262                                          0.91%    failed      0.02%\n",
      "  32  Q8K2B3.1_Mus_musculus                                                        0.00%    passed     71.37%\n",
      "  33  Q9KPA4.2_Vibrio_cholerae_O1_biovar_El_Tor_str._N16961                        0.00%    passed     80.87%\n",
      "  34  Q8ZQU3.1_Salmonella_enterica_subsp._enterica_serovar_Typhimurium_str._LT2    0.00%    passed     99.24%\n",
      "  35  Q94AY1.1_Arabidopsis_thaliana                                                0.23%    passed     83.85%\n",
      "  36  Q0QF01.2_Sus_scrofa                                                          0.00%    passed     58.92%\n",
      "  37  Q972D2.1_Sulfurisphaera_tokodaii_str._7                                      2.27%    failed      0.00%\n",
      "  38  Q59661.1_Paracoccus_denitrificans                                            0.00%    passed     91.15%\n",
      "  39  P08065.4_Bacillus_subtilis_subsp._subtilis_str._168                          0.00%    passed     16.24%\n",
      "  40  Q8ZMX9.1_Salmonella_enterica_subsp._enterica_serovar_Typhimurium_str._LT2    0.00%    passed     64.21%\n",
      "  41  Q9YHT1.2_Gallus_gallus                                                       0.00%    passed     58.04%\n",
      "  42  Q9K107.1_Neisseria_meningitidis_MC58                                         0.68%    failed      0.00%\n",
      "  43  Q7ZVF3.2_Danio_rerio                                                         0.00%    passed     97.19%\n",
      "  44  O66973.1_Aquifex_aeolicus_VF5                                                0.91%    failed      0.00%\n",
      "  45  P64175.1_Mycobacterium_tuberculosis_variant_bovis_AF2122/97                  0.23%    passed     78.31%\n",
      "  46  Q51363.1_Pseudomonas_aeruginosa_PAO1                                         0.00%    passed     69.26%\n",
      "  47  Q8U8J4.1_Agrobacterium_fabrum_str._C58                                       0.23%    failed      0.00%\n",
      "  48  P20922.1_Proteus_vulgaris                                                    0.00%    passed     91.92%\n",
      "  49  P83223.2_Shewanella_oneidensis_MR-1                                         15.87%    failed      0.03%\n",
      "  50  Q49617.2_Mycobacterium_leprae_TN                                             0.45%    failed      0.00%\n",
      "  51  Q9A4C3.1_Caulobacter_vibrioides_CB15                                         0.00%    failed      0.00%\n",
      "  52  Q9KDJ5.1_Alkalihalobacillus_halodurans_C-125                                 2.27%    passed     73.74%\n",
      "  53  Q8ZD80.2_Yersinia_pestis                                                     0.00%    passed     95.98%\n",
      "  54  Q60356.1_Methanocaldococcus_jannaschii_DSM_2661                              1.13%    failed      0.00%\n",
      "  55  Q09508.3_Caenorhabditis_elegans                                              0.00%    passed     94.05%\n",
      "  56  Q9Z4P0.1_Shewanella_frigidimarina_NCIMB_400                                 15.87%    failed      0.00%\n",
      "  57  D9PU00.1_Methanothermobacter_marburgensis_str._Marburg                       0.68%    failed      2.35%\n",
      "  58  P0AC41.1_Escherichia_coli_K-12                                               0.00%    passed     99.24%\n",
      "  59  P74562.1_Synechocystis_sp._PCC_6803_substr._Kazusa                           0.00%    failed      0.35%\n",
      "  60  Q9U3X4.1_Dictyostelium_discoideum                                            0.00%    passed     99.61%\n",
      "  61  Q00711.1_Saccharomyces_cerevisiae_S288C                                      0.00%    passed     99.81%\n",
      "  62  Q98AV8.1_Mesorhizobium_japonicum_MAFF_303099                                 0.23%    failed      0.03%\n",
      "  63  Q801S2.1_Xenopus_laevis                                                      0.00%    passed     80.81%\n",
      "  64  P47052.1_Saccharomyces_cerevisiae_S288C                                      0.00%    passed     99.69%\n",
      "  65  Q9ZMP0.1_Helicobacter_pylori_J99                                             0.00%    passed     43.53%\n",
      "  66  P44894.1_Haemophilus_influenzae_Rd_KW20                                      0.00%    passed     65.10%\n",
      "  67  P65500.1_Mycobacterium_tuberculosis_variant_bovis_AF2122/97                  0.45%    failed      0.00%\n",
      "  68  Q8XA23.1_Escherichia_coli_O157                                               0.00%    passed     89.79%\n",
      "  69  G4V4G6.1_Serratia_sp._ATCC_39006                                             0.00%    passed     99.32%\n",
      "  70  Q5R616.1_Pongo_abelii                                                        0.00%    passed     76.00%\n",
      "  71  Q97K95.1_Clostridium_acetobutylicum_ATCC_824                                 8.62%    failed      0.00%\n",
      "  72  Q9ZPX5.1_Arabidopsis_thaliana                                                0.00%    passed     97.66%\n",
      "  73  P31039.3_Bos_taurus                                                          0.00%    passed     49.04%\n",
      "  74  Q7M827.1_Wolinella_succinogenes_DSM_1740                                     0.00%    passed     88.18%\n",
      "  75  O06913.2_Helicobacter_pylori_26695                                           0.00%    passed     52.45%\n",
      "  76  Q94523.3_Drosophila_melanogaster                                             0.00%    passed     99.23%\n",
      "  77  Q9UTJ7.1_Schizosaccharomyces_pombe_972h-                                     0.00%    passed     99.54%\n",
      "  78  P32614.1_Saccharomyces_cerevisiae_S288C                                     16.10%    failed      0.10%\n",
      "  79  Q07WU7.2_Shewanella_frigidimarina_NCIMB_400                                 15.87%    failed      0.00%\n",
      "  80  P00363.3_Escherichia_coli_K-12                                               0.00%    passed     65.78%\n",
      "  81  Q28ED0.1_Xenopus_tropicalis                                                  0.00%    passed     83.44%\n",
      "  82  Q8Z4K0.1_Salmonella_enterica_subsp._enterica_serovar_Typhi                   0.00%    passed     60.57%\n",
      "  83  P51054.1_Coxiella_burnetii_RSA_493                                           0.23%    passed     86.02%\n",
      "  84  Q8XWM7.1_Ralstonia_solanacearum_GMI1000                                      0.23%    passed     45.63%\n",
      "  85  Q1RHB9.1_Rickettsia_bellii_RML369-C                                          0.00%    passed     99.97%\n",
      "  86  Q8HXW3.1_Macaca_fascicularis                                                 0.00%    passed     76.79%\n",
      "****  TOTAL                                                                        1.40%  30 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "NOTE: Q5R616.1_Pongo_abelii is identical to P31040.2_Homo_sapiens but kept for subsequent analysis\n",
      "Creating fast initial parsimony tree by random order stepwise addition...\n",
      "0.026 seconds, parsimony score: 7737 (based on 415 sites)\n",
      "Generating 1000 samples for ultrafast bootstrap (seed: 393519)...\n",
      "\n",
      "NOTE: 7 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -39398.388\n",
      "Optimal log-likelihood: -39123.141\n",
      "Parameters optimization took 1 rounds (0.034 sec)\n",
      "Computing ML distances based on estimated model parameters... 0.262 sec\n",
      "Computing BIONJ tree...\n",
      "0.003 seconds\n",
      "Log-likelihood of BIONJ tree: -38975.566\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 0.257 second\n",
      "Computing log-likelihood of 98 initial trees ... 0.832 seconds\n",
      "Current best score: -38975.566\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -38943.254\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 2: -38939.209\n",
      "Iteration 10 / LogL: -38948.100 / Time: 0h:0m:2s\n",
      "Iteration 20 / LogL: -38972.367 / Time: 0h:0m:4s\n",
      "Finish initializing candidate tree set (20)\n",
      "Current best tree score: -38939.209 / CPU time: 3.990\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -38940.721 / Time: 0h:0m:6s (0h:0m:15s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 39: -38938.928\n",
      "Iteration 40 / LogL: -38942.261 / Time: 0h:0m:7s (0h:0m:20s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 43: -38938.735\n",
      "Iteration 50 / LogL: -38940.806 / Time: 0h:0m:9s (0h:0m:18s left)\n",
      "Log-likelihood cutoff on original alignment: -38984.823\n",
      "UPDATE BEST LOG-LIKELIHOOD: -38938.735\n",
      "Iteration 60 / LogL: -38939.619 / Time: 0h:0m:11s (0h:0m:16s left)\n",
      "Iteration 70 / LogL: -38940.868 / Time: 0h:0m:13s (0h:0m:14s left)\n",
      "Iteration 80 / LogL: -38945.117 / Time: 0h:0m:15s (0h:0m:12s left)\n",
      "Iteration 90 / LogL: -38941.785 / Time: 0h:0m:17s (0h:0m:10s left)\n",
      "Iteration 100 / LogL: -38941.987 / Time: 0h:0m:19s (0h:0m:8s left)\n",
      "Log-likelihood cutoff on original alignment: -38982.988\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.999\n",
      "UPDATE BEST LOG-LIKELIHOOD: -38938.735\n",
      "Iteration 110 / LogL: -38940.970 / Time: 0h:0m:21s (0h:0m:17s left)\n",
      "Iteration 120 / LogL: -38947.194 / Time: 0h:0m:23s (0h:0m:15s left)\n",
      "Iteration 130 / LogL: -38939.215 / Time: 0h:0m:25s (0h:0m:13s left)\n",
      "Iteration 140 / LogL: -38940.615 / Time: 0h:0m:27s (0h:0m:11s left)\n",
      "TREE SEARCH COMPLETED AFTER 144 ITERATIONS / Time: 0h:0m:28s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -38938.735\n",
      "Optimal log-likelihood: -38938.735\n",
      "Parameters optimization took 1 rounds (0.015 sec)\n",
      "BEST SCORE FOUND : -38938.735\n",
      "Creating bootstrap support values...\n",
      "Split supports printed to NEXUS file SDHA_ncbi_muscle_trim.splits.nex\n",
      "Total tree length: 25.084\n",
      "\n",
      "Total number of iterations: 144\n",
      "CPU time used for tree search: 164.323 sec (0h:2m:44s)\n",
      "Wall-clock time used for tree search: 27.886 sec (0h:0m:27s)\n",
      "Total CPU time used: 165.640 sec (0h:2m:45s)\n",
      "Total wall-clock time used: 28.225 sec (0h:0m:28s)\n",
      "\n",
      "Computing bootstrap consensus tree...\n",
      "Reading input file SDHA_ncbi_muscle_trim.splits.nex...\n",
      "86 taxa and 330 splits.\n",
      "Consensus tree written to SDHA_ncbi_muscle_trim.contree\n",
      "Reading input trees file SDHA_ncbi_muscle_trim.contree\n",
      "Log-likelihood of consensus tree: -38940.278\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                SDHA_ncbi_muscle_trim.iqtree\n",
      "  Maximum-likelihood tree:       SDHA_ncbi_muscle_trim.treefile\n",
      "  Likelihood distances:          SDHA_ncbi_muscle_trim.mldist\n",
      "\n",
      "Ultrafast bootstrap approximation results written to:\n",
      "  Split support values:          SDHA_ncbi_muscle_trim.splits.nex\n",
      "  Consensus tree:                SDHA_ncbi_muscle_trim.contree\n",
      "  Screen log file:               SDHA_ncbi_muscle_trim.log\n",
      "\n",
      "Date and Time: Thu Sep 29 11:10:29 2022\n"
     ]
    }
   ],
   "source": [
    "## Run the tree making tool \n",
    "\n",
    "!iqtree -s SDHA_ncbi_muscle_trim.fasta -m LG -B 1000 -nt 6 -pre SDHA_ncbi_muscle_trim \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab36df8c-6db6-4c37-a5bc-146f0bc3b962",
   "metadata": {},
   "outputs": [],
   "source": [
    "## if it does not work, try \n",
    "iqtree = \"/home/julieb/anaconda3/bin/iqtree\" # write here the path of your iqtree install\n",
    "in_file = \"SDHA_ncbi_muscle_trim.fasta\"\n",
    "pre = \"SDHA_ncbi_muscle_trim\"\n",
    "model = \"LG,WAG\"\n",
    "rate = \"G,I\"\n",
    "bt = \"1000\"\n",
    "iqtree_result=subprocess.run([iqtree,\"-s\", in_file, \"-mset\", model, \"-mrate\", rate, \"-bb\", bt, \"-pre\", out_file])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57a26952-d409-484c-8bfa-744ab4b93541",
   "metadata": {},
   "outputs": [],
   "source": [
    "# install figtree \n",
    "# in a console, type\n",
    "# conda install -c bioconda figtree\n",
    "\n",
    "# run figtree from a consome\n",
    "# figtree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e76f912-283c-41c3-9821-d4ca1827236b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if you struggle with figtree install, you can view your .contree file using Jalview\n",
    "# in the mafft window, choose File > Load associated tree"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
